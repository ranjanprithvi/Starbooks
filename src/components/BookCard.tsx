import React from "react";
import { Book } from "../services/book-service";
import { Card, CardBody, HStack, Heading, Image, Text } from "@chakra-ui/react";
import Rating from "./Rating";

interface Props {
    book: Book;
}

const BookCard = ({ book }: Props) => {
    return (
        <Card
            height={"100%"}
            borderRadius={10}
            overflow="hidden"
            boxShadow={"2xl"}
        >
            <Image
                src={
                    book.coverImage ||
                    "https://bookcart.azurewebsites.net/Upload/Default_image.jpg"
                }
                borderRadius={10}
                height="75%"
                // overflow={"clip"}
            />
            <CardBody>
                <Heading marginY="5px" fontSize="xl">
                    {book.title +
                        (book.yearPublished ? ` (${book.yearPublished})` : "")}
                </Heading>
                <HStack justifyContent={"space-between"}>
                    <Text color={"gray.500"}>{book.author.name}</Text>
                    <Rating rating={book.rating} />
                </HStack>
            </CardBody>
        </Card>
    );
};

export default BookCard;
