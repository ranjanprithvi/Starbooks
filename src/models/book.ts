import { Sort } from "../components/common/SortSelector";
import { Author } from "./author";
import { Genre } from "./genre";

export interface Book {
    _id: string;
    title: string;
    coverImage: string;
    author: Author;
    genre: Genre;
    rating: number;
    yearPublished: number;
    description: string;
    numberInStock: number;
}

export interface BookQuery {
    author?: string;
    genre?: string;
    search?: string;
    sortBy?: string;
    populate?: string;
}

export const defaultBookCover = "/src/assets/default-no-cover.jpeg";

export const bookSortFields: Sort[] = [
    { name: "Date Added", value: "-_id" },
    { name: "Title", value: "title" },
    { name: "Rating", value: "-rating" },
    { name: "Year Published", value: "-yearPublished" },
];
