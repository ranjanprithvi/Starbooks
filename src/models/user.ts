export interface User {
    _id: string;
    name: string;
    email: string;
    membershipExpiry?: Date;
    isAdmin: boolean;
}

export interface UserQuery {
    search?: string;
    sortBy?: string;
    populate?: string;
}
