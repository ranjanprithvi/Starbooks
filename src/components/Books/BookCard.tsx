import { Card, CardBody, HStack, Heading, Image, Text } from "@chakra-ui/react";
import Rating from "./Rating";
import { Book, defaultBookCover } from "../../models/book";

interface Props {
    book: Book;
}

const BookCard = ({ book }: Props) => {
    return (
        <Card
            height={"100%"}
            borderRadius={10}
            overflow="hidden"
            boxShadow={"2xl"}
        >
            <Image
                src={book.coverImage || defaultBookCover}
                borderRadius={10}
                height="75%"
                // overflow={"clip"}
            />
            <CardBody>
                <Heading marginY="5px" fontSize="xl">
                    {book.title +
                        (book.yearPublished ? ` (${book.yearPublished})` : "")}
                </Heading>
                <HStack justifyContent={"space-between"}>
                    <Text color={"gray.500"}>{book.author.name}</Text>
                    <Rating rating={book.rating} />
                </HStack>
            </CardBody>
        </Card>
    );
};

export default BookCard;
