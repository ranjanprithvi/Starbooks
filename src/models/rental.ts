import { Book } from "./book";
import { User } from "./user";

export interface Rental {
    _id: string;
    book: Book;
    user: User;
    dateReturned?: Date;
}
