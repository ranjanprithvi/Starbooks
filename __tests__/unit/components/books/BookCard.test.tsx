import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import BookCard from "../../../../src/components/Books/BookCard";
import { defaultBookCover } from "../../../../src/models/book";

describe("BookCard", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(
                <BookCard
                    book={{
                        _id: "123",
                        title: "test",
                        author: { _id: "1", name: "author1" },
                        genre: { _id: "1", name: "genre1" },
                        yearPublished: 2021,
                        rating: 5,
                        coverImage:
                            "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1681496408i/58536026.jpg",
                        description: "test",
                        numberInStock: 1,
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders correctly with no cover image", () => {
        render(
            <BookCard
                book={{
                    _id: "123",
                    title: "test",
                    author: { _id: "1", name: "author1" },
                    genre: { _id: "1", name: "genre1" },
                    yearPublished: 2021,
                    coverImage: "",
                    rating: 5,
                    description: "test",
                    numberInStock: 1,
                }}
            />
        );
        expect(screen.getByRole("img").getAttribute("src")).toBe(
            defaultBookCover
        );
    });
});
