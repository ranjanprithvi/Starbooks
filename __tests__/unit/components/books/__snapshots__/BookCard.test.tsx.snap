// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard renders correctly 1`] = `
<div
  className="chakra-card css-12r5yzq"
>
  <img
    className="chakra-image css-1ec1yfo"
    src="https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1681496408i/58536026.jpg"
  />
  <div
    className="chakra-card__body css-0"
  >
    <h2
      className="chakra-heading css-119ujjf"
    >
      test (2021)
    </h2>
    <div
      className="chakra-stack css-1y3f6ad"
    >
      <p
        className="chakra-text css-ax9bdp"
      >
        author1
      </p>
      <span
        className="chakra-badge css-4z4zi4"
      >
        5
      </span>
    </div>
  </div>
</div>
`;
